proxy/src/main/java/com/velocitypowered/proxy/connection/MinecraftConnection.java

```diff
@@ -99,20 +153,76 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
 
     @Override
     public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
+        if (pluginsSessionHandlers != null && !pluginsSessionHandlers.isEmpty()) {
+            synchronized (pluginsSessionHandlers) {
+                for (MinecraftSessionHandler handler : pluginsSessionHandlers) {
+                    handler.writabilityChanged();
+                }
+            }
+        }
         if (sessionHandler != null) {
             sessionHandler.writabilityChanged();
         }
     }
 
+    public void addCustomSessionHandler(MinecraftSessionHandler handler) {
+        if (this.pluginsSessionHandlers == null) {
+            this.pluginsSessionHandlers = Collections.synchronizedList(new ArrayList<>());
```

Comment:
I would probably prefer using `CopyOnWriteArrayList` here. It is perfect since plugins are unlikely to register and unregister listeners frequently, but iterations occur often.